Signup: Use Redux for handling/persisting signup progress

jsnmoon opened the pull request:
Part of step 2 for #6709; builds on #25645. See p9xfpQ-8z-p2 for further discussion.

This change reworks the SignupProgressStore to use Redux instead of Flux. The SignupProgressStore is responsible for tracking and persisting the user's progression through our various signup flows.

Previously, this store persisted the current user's signup progress into localStorage via the [store.js](https://www.npmjs.com/package/store). It now relies on Calypso's [Redux state persistence mechanism](https://github.com/Automattic/wp-calypso/blob/master/client/state/initial-state.js#L166) which relies on [localforage](https://www.npmjs.com/package/localforage).

Existing signup flow behavior should remain identical as before.

Some notes: 

1. The signup flow is designed to cross the logged-out/logged-in boundary. Since the store persisted in logged-out state isn't loaded in the logged-in state, I'm assuming that the progress store is only useful during the signup process to the logged-out user. Once a user completes signup by completing the `user` step, we discard the logged-out state.

2. Upon construction of the signup flow controller, we invoke a function to discard steps not associated with the current flow. This ensures that we clear out stale/invalid steps persisted by the progress state in the Redux store.

# Testing instructions
1. Spin up this branch either on the [live branch](https://calypso.live/?branch=reduxify/signup-progress-store-part-2) or locally.
2. Go through the entirety of a signup flow such as [`/start`](http://calypso.localhost:3000/start/domains). 
3. Verify that you are correctly redirected following the signup flow. Depending on your A/B test configuration, this is either your new site or the new user checklist page.

Commit:
Signup: Fix linting errors in ProgressStore

Commit:
Signup: Rename changeSignupFlow parameter name

Use explicit parameter name `flowName` instead of `flow`, which could be referring to an entire `flow` object.

Commit:
Signup: Modularize mock file for flow config

Commit:
Signup: Reduxify the progress store

Commit:
Signup: Update dependency store test re: redux

Ensure that the redux store is being set for the newly reduxified progress store.

Commit:
Signup: Reduxify the flow controller

This removes usage of `store` for progress persistence.

Commit:
Signup: Use reduxified signup progress state

alisterscott commented:
I'm looking at why the e2e tests are failing, it looks like upon completing sign up on this branch the customer is redirected straight into /wp-admin on the new site. Any reason for this?

jsnmoon commented:
> I'm looking at why the e2e tests are failing, it looks like upon completing sign up on this branch the customer is redirected straight into /wp-admin on the new site. Any reason for this? (@alisterscott)

Hmm, it looks like the refactor unintentionally affected the redirection logic. I'll update when I have something more concrete to report.

alisterscott commented:
I see a console error revisiting the start page after going through the flow:

```
build.382083f4ff53ca58eb95.opt.js:1 Error: Tried to read redux state before store was ready
    at Object.get (signup.7fdc90687b2b92753b37.opt.js:1)
```

Steps to Reproduce:

1. Visit https://calypso.live/start?branch=reduxify/signup-progress-store-part-2 in a new browser window (logged out)
2. Progress to step 4 account
3. Revisit https://calypso.live/start?branch=reduxify/signup-progress-store-part-2
4. Nav back through steps to start page (about)
5. Revisit https://calypso.live/start?branch=reduxify/signup-progress-store-part-2
6. Error comes up in console and screen won't load - see screen recording:

![sign up](https://user-images.githubusercontent.com/128826/42503374-d5ed72b2-847b-11e8-9578-9c0fe25d7651.gif)

alisterscott reviewed (dismissed):
I left a comment with an error that occurs when revisiting start

alisterscott commented:
Re-tested session store bug and works well now 👍

alisterscott reviewed (approved):
Looks good from a product perspective.
All sign up e2e tests passing.
Couldn't find any issues in manual testing.
Tested IE11 sign ups also.
Thanks for this change.

blowery commented:
@jsnmoon looks like some changes happened to signup around this PR. Could you take a look?

jsnmoon commented:
> @jsnmoon this needs a minor rebase and then I think we can land it. Can you take a look? I nearly committed the rebase myself, but I'm not 100% sure it's as simple as it appears. (@blowery)

This has been rebased to account for changes introduced by #25690.

alisterscott commented:
Hi @jsnmoon 

I retested this since rebasing and I can't see @taggon's fix in this branch any more.

**Steps**

1. As a non-logged-in customer visit https://calypso.live/start?branch=reduxify/signup-progress-store-part-2
2. Proceed to the account screen
3. Enter email of existing account
4. Use the log in now link
5. Customer is redirected back to the start of the sign up flow when it should keep their position

Commit:
Signup: Manually revert #25690 to prune code

Currently, reduxifying the signup flow renders #25690 ineffective. This is due to the fact that we currently don't have a way to persist Redux state between the logged-out/logged-in boundary.

jsnmoon commented:
> I retested this since rebasing and I can't see @taggon's fix in this branch anymore. (@alisterscott)

Unfortunately, we currently don't have a way to persist Redux state between the logged-out/logged-in boundary. This sadly means we would intentionally undo #25690 with this PR.

I currently don't have a straightforward fix for this; I started this reduxification effort assuming that we didn't have significant UX relying on passing data through the logged-in boundary. 

Sorry @taggon :(

jsnmoon commented:
Unrelated pages are currently failing on `wp-e2e-tests-full` at the moment. Should be safe to disregard.

blowery commented on client/lib/signup/progress-store.js:
```diff
@@ -4,139 +4,85 @@
+
+	get() {
+		if ( ! this.reduxStore ) {
+			throw new Error( 'Tried to read redux state before store was ready' );
```
maybe just return instead of throwing?

blowery commented on client/lib/signup/progress-store.js:
```diff
@@ -4,139 +4,85 @@
+		if ( ! this.reduxStore ) {
+			throw new Error( 'Tried to read redux state before store was ready' );
+		}
+		return cloneDeep( getSignupProgress( this.reduxStore.getState() ) );
```
why the `cloneDeep`?

blowery replied:
it was more curiosity than anything. I assume this code is going to get thrown out anyway when we remove the flux store, so ... probably fine to leave it. `cloneDeep` isn't cheap.

jsnmoon replied:
I just reviewed all instances of `ProgressStore.get()` and confirmed that we don't mutate the object anymore with this change. I'll go ahead and remove it.

blowery commented on client/lib/signup/progress-store.js:
```diff
@@ -4,139 +4,85 @@
+			steps: input,
+		} );
+	},
+	setReduxStore( reduxStore ) {
```
I think you can use the `redux-bridge` instead of having to hold onto a ref to the redux store.

jsnmoon replied:
It seems like we'd want to continue using the `setReduxStore` pattern in the signup (for now) to preserve consistency with other signup stores. 

Also, it seems like value change subscription pattern is not yet supported in `redux-bridge`, so it might not be sufficient here without additional work (e.g. `SignupProgressStore.on( 'change', callbackFn );`). I might be wrong, though.

blowery commented on client/signup/main.jsx:
```diff
@@ -88,23 +92,21 @@ class Signup extends React.Component {
 	}
 
-	componentWillMount() {
+	UNSAFE_componentWillMount() {
```
can we move this to DidMount or the constructor instead of moving to UNSAFE_*?

jsnmoon replied:
Given the large and interconnected nature of the main signup script, I wanted to avoid introducing more changes than necessary with this PR.

It's certainly worth considering once we land this change, though.

blowery commented on client/signup/main.jsx:
```diff
@@ -291,12 +286,11 @@ class Signup extends React.Component {
 
 	handleFlowComplete = ( dependencies, destination ) => {
-		debug( 'The flow is completed. Logging you in...' );
+		debug( `The flow is completed. Destination: ${ destination }` );
```
super minor, but using a string placeholder avoids the string concat when debug is disabled. like
`debug( 'The flow is completed. Destination: %s', destination )`
